Mark `useDispatch` as deprecated

leon-vogt opened the pull request:
closes #206 

Is the deprecation message good enough?
```
useDispatch() is a deprecated mixin from Stimulus-Use. Please use the built-in dispatch() function from Stimulus: https://stimulus.hotwired.dev/reference/controllers#cross-controller-coordination-with-events
```

Two questions:
1. Should other docs also be updated?
- docs/application-controller.md
- docs/application.md
- docs/debug.md
2. Should a "deprecation" file be created, to track the deprecation cycle?

marcoroth commented on docs/use-dispatch.md:
```diff
@@ -2,6 +2,8 @@
 
 Adds a `dispatch` helper function to emit custom events. Useful to communicate between different controllers.
 
+**Deprecated** : useDispatch() is a deprecated mixin from Stimulus-Use. Please use the built-in dispatch() function from Stimulus: https://stimulus.hotwired.dev/reference/controllers#cross-controller-coordination-with-events
```
```suggestion
**Deprecated**: `useDispatch()` is deprecated. Please use the built-in `this.dispatch()` function from Stimulus: https://stimulus.hotwired.dev/reference/controllers#cross-controller-coordination-with-events
```

marcoroth commented on src/stimulus-use.ts:
```diff
@@ -55,6 +55,10 @@ export class StimulusUse {
     this.logger.groupEnd()
   }
 
+  warn = (functionName: string, args: any): void => {
```
I think you can leave out the function name here and remove it from the `warn` function signature. It should anyway be visible from the backtrace itself.

marcoroth commented on src/use-dispatch/use-dispatch.ts:
```diff
@@ -49,6 +49,7 @@ export class UseDispatch extends StimulusUse {
     // dispatch the event from the given element or by default from the root element of the controller
     targetElement.dispatchEvent(event)
 
+    warn('dispatch', 'useDispatch() is a deprecated mixin from Stimulus-Use. Please use the built-in dispatch() function from Stimulus: https://stimulus.hotwired.dev/reference/controllers#cross-controller-coordination-with-events');
```
This doesn't seem to work. 

Ironically if you have this controller here:

```js
import { Controller } from '@hotwired/stimulus'
import { useDispatch } from 'stimulus-use'

export default class extends Controller {
  connect() {
    useDispatch(this)
  }

  add() {
    this.dispatch('add', { quantity: 1 })
  }
}
```
It's already using the `this.dispatch()` function from Stimulus itself. So this code in our `dispatch` function is actually never called again. 

Which is why we should move the deprecation warning into our `useDispatch()` function itself.

leon-vogt replied:
Hm, that's odd...
When I run your controller example with two additional console.log outputs:

```js
import { Controller } from '@hotwired/stimulus'
import { useDispatch } from 'stimulus-use'

export default class extends Controller {
  connect() {
    console.log("ITEM Controller connected");
    useDispatch(this)
  }

  add() {
    console.log("ITEM Controller#add was called");
    this.dispatch('add', { quantity: 1 })
  }
}
```

I get following Output in the Console:

![image](https://user-images.githubusercontent.com/17851143/197958871-7b2dc8e3-638c-4f61-86a5-91239d29c837.png)


---

The Reason why I didn't put the deprecation warning into the `useDispatch()` function itself is, it would print the warning on every request, even though I don't use or import the `useDispatch()` function.
That's why I put the deprecation warning in the `dispatch()` call.

Do you thinks that's a bad idea?

marcoroth reviewed (changes requested):
Hey @leon-vogt, thanks for taking this on and opening this pull requests! 🤗 

> 1. Should other docs also be updated?

Yes, I think that would make sense!

> 2. Should a "deprecation" file be created, to track the deprecation cycle?

I don't think we need that (yet). It should be fine to have the warning in there. And once we remove the function itself we can also get rid of the deprecation warning with it.

I left some comments, but this looks good to me otherwise! 🙌🏼

leon-vogt commented:
@marcoroth thanks for taking the time and review the code!
I will update the other docs as well and adopt your suggestions

leon-vogt commented:
The deprecation warning is now in the documentation visible and looks like this:
![image](https://user-images.githubusercontent.com/17851143/198351030-68df75ad-5088-4633-af25-d47ca0f95b7e.png)

I removed the `useDispatch` example from `docs/debug.md` (To propose not outdated things)
And added deprecation warnings in the following files:
- docs/application-controller.md
- docs/application.md
- docs/use-dispatch.md

--- 

The one thing I wasn't sure about, was the [cart example](https://stimulus-use.github.io/stimulus-use/#/application-controller?id=example-building-a-cart-counter-with-the-dispatch-helper)...
Maybe an example with another mixin like `isPreview` would make more sense now.

marcoroth commented:
Thank you, this looks very good now!

> The one thing I wasn't sure about, was the [cart example](https://stimulus-use.github.io/stimulus-use/#/application-controller?id=example-building-a-cart-counter-with-the-dispatch-helper)... Maybe an example with another mixin like `isPreview` would make more sense now.

You are right. Either we remove it, or we just keep it in there to actually have a demo for the `dispatch` function in general. But if you ask me we can just keep it there for now.

marcoroth reviewed (approved):
Awesome, thank you @leon-vogt for pulling this one off! 🎉
